src/Lookup/ClosestReferencedEntityIdLookupException.php

```diff
@@ -0,0 +1,65 @@
+<?php
+
+namespace Wikibase\DataModel\Services\Lookup;
+
+use Exception;
+use RuntimeException;
+use Wikibase\DataModel\Entity\EntityId;
+use Wikibase\DataModel\Entity\PropertyId;
+
+/**
+ * @since 3.10
+ *
+ * @license GPL-2.0-or-later
+ * @author Marius Hoch
+ */
+class ClosestReferencedEntityIdLookupException extends RuntimeException {
+
+	/**
+	 * @var EntityId
+	 */
+	private $fromId;
+
+	/**
+	 * @var PropertyId
+	 */
+	private $propertyId;
+
+	/**
+	 * @var EntityId[]
+	 */
+	private $toIds;
+
+	/**
+	 * @param EntityId $fromId
+	 * @param PropertyId $propertyId
+	 * @param EntityId[] $toIds
+	 * @param string|null $message
+	 * @param Exception|null $previous
+	 */
+	public function __construct(
+		EntityId $fromId,
+		PropertyId $propertyId,
+		array $toIds,
+		$message = null,
+		Exception $previous = null
+	) {
+		$this->fromId = $fromId;
+		$this->propertyId = $propertyId;
+		$this->toIds = $toIds;
+
+		$targets = array_map(
+			function ( EntityId $entityId ) {
+				return $entityId->getSerialization();
+			},
+			$toIds
+		);
+		$targets = implode( ', ', $targets );
+
+		$message = $message ?: 'Closest referenced entity id lookup failed. Tried to find the closest entity out of ' .
+			$targets . ' linked from ' . $fromId->getSerialization() . ' via ' . $propertyId->getSerialization();
+
```

Comment:
No objection to this, but I do wonder if it is really needed to construct such an elaborate message. Can't the info be kept in a more structured form? Perhaps when the message construction is needed it can be done on demand in the accessing method (not sure how this usually works for exceptions). That way you avoid the work in the constructor.

---

I'd extract a `buildMessage` and perhaps a `buildTargetsString` methods.